homework/hw_w01.py

```diff
@@ -0,0 +1,210 @@
+        query_columns += f"{key}, "
+        # Use format value to handle value types in SQL
+        variable_list.append(row[key])
+    values = "%s, " * len(row.keys())
+    # Index to -2 to remove ", " from last value
+    query = f"INSERT INTO {table} ({query_columns[:-2]}) VALUES ({values[:-2]})"
+    return query, variable_list
+
+def load_table(conn, table_name, frame):
+    """Load a dataframe into a specified table"""
+    # Build table
+    for index, row in frame.iterrows():
+        query_str, var_tuple = build_query(row, table_name, "INSERT")
+        #print(query_str)
+        conn.insert_query(query_str, var_tuple)
+
+def _torch_tables(conn, tables=['products', 'orders',
+                                'customers', 'employees', 'offices',  ]):
+    """Delete the contents of the database"""
+    # This is for testing only, method should not be used.
+    for table in tables:
+        conn._delete_query(f"DELETE FROM {table};")
+
+def load_products_file(conn, filename):
+    """Load the records from the products file"""
+    products = pd.read_csv(filename)
+    # From products, take all columns
+    load_table(conn, 'products', products)
+
+def load_orders_file(conn, filename):
+    """Load the records from the orders file"""
+    orders_full = pd.read_csv(filename)
+    # Convert byte columns to string literals
+    orders_full["customer_location"] = (orders_full["customer_location"]
+        .apply(ast.literal_eval))
+    # Fix nulls to None as it messes with the dates
+    orders_full = orders_full.where(pd.notnull(orders_full), None)
+    # Split orders into three tables; customers, orders and order_items
+    load_customers_table(conn, orders_full)
+    #Second part, orders
+    load_orders_table(conn, orders_full)
+    # Third part, order_items
+    load_order_items_table(conn, orders_full)
+
+def load_customers_table(conn, orders_full):
+    """Load the records into the customers_table"""
+    customers_col_list = ['customer_number', 'customer_name',
+                          'contact_last_name', 'contact_first_name',
+                          'city', 'state', 'country',
+                          'sales_rep_employee_number', 'credit_limit',
+                          'customer_location']
+    customers_df = orders_full[customers_col_list].drop_duplicates()
+    load_table(conn, 'customers', customers_df)
+
+def load_orders_table(conn, orders_full):
+    """Load the records into the orders table"""
+    orders_col_list = [ 'order_number', 'customer_number', 'order_date',
+                       'required_date', 'shipped_date', 'status',
+                       'comments' ]
+    orders_df = orders_full[orders_col_list].drop_duplicates()
+    load_table(conn, 'orders', orders_df)
+
+def load_order_items_table(conn, orders_full):
+    """Load the records into the order_items table"""
+    # Create new index for order_item that combines the order_number with the
```

Comment:
As we mentioned in class -- this is an anti-patter, because SQL gives you an official way to create an index that is the combination of two columns (composite primary key) WITHOUT needing to copy the data to a new, mangled column!